spec/rspec/rails/example/controller_example_group_spec.rb

```diff
@@ -19,6 +19,13 @@ def group_for(klass)
       expect(group.included_modules).to include(RSpec::Rails::Matchers::RoutingMatchers)
     end
 
+    it "handles dynamic matchers with keywords (big change from ruby 2 to ruby 3)" do
+      example = group.new
+      example.define_singleton_method(:has_val?) { |val:| val == 1 }
```

Comment:
The code comment to `method_missing` claims:
>      # If method is a named_route, delegates to the RouteSet associated with
>      # this controller.

Do you think this statement still stands?
You mention dynamic matchers, what is the real case scenario that failed for you?

Reply:
> Do you think this statement still stands?

Yeah, that's what the `if` statement does ([here](https://github.com/rspec/rspec-rails/blob/e4b3953561ae5193459685ec009e91ae3705debb/lib/rspec/rails/example/controller_example_group.rb#L173-L174)). The method did not intend to change the arguments, and in older versions of Ruby, it didn't. But with the new keyword change, this method accidentally moved the keyword arguments into the ordinal arguments. In Ruby 3, the way to do what this method used to do is to accept an ellipsis instead of an asterisk. Or to accept an asterisk for the ordinals and two asterisks for the keywords. But you can't use those if you also need to support Ruby 2.

Anyway here's an example showing how the intermediate `method_missing` accidentally changes the structure of the arguments:

```sh
# You *must* be on Ruby 3 to see the issue
ruby -v -W0 -e '
 # The real method_missing, this looks at the name, realizes it matches the `has_whatever`
 # matcher, and constructs / returns RSpec::Matchers::Has, out of the arguments
 def method_missing(name, a:)
   "SUCCESS! `#{name}(a: #{a.inspect})` was invoked"
 end

 # ControllerExampleGroup\'s method_missing. This checks if it should do stuff
 # related to route set (or whatever), decides it shouldn\'t, and passes it up
 # to the overwritten `method_missing`, thinking it has had no impact, but in
 # reality, it moved the keyword arguments into the ordinal arguments
 def self.method_missing(*)
   super
 end

 def self.HEADER(h) = puts("\n\e[33m#{h}\e[0m")

 HEADER "Invocation will fail because intermediate method_missing changes args"
 have_failure a: 111111111 rescue p $!

 HEADER "Now we will fix the args"
 singleton_class.class_eval { ruby2_keywords :method_missing }

 HEADER "Now invocation will succeed"
 puts have_success a: 222222222
'
```

Output:
<img width="869" alt="image" src="https://user-images.githubusercontent.com/77495/125864948-23be9fef-ec51-4d1d-98b6-09933b94bfcf.png">


> You mention dynamic matchers, what is the real case scenario that failed for you?

The same as what is shown. I had a helper object that made assertions, people thought it wasn't "RSpecy" enough, so I changed `obj.assert_whatever arg: val` to `expect(obj).to have_whatever(arg: val)`, which failed in the controller spec, because the controller spec included this module.

Here is a standalone example, I tried to make it slightly more realistic, since my test apparently didn't communicate it. Run it in Ruby 3 and it will fail, uncomment line 7 and it will pass:

```ruby
require 'bundler/inline'
gemfile do
  source 'https://rubygems.org'
  gem 'actionpack', '= 6.1.4', require: 'action_controller/railtie'
  gem 'rspec-rails', '= 5.0.1', require: 'rspec/rails'
end

Class.new Rails::Application

module RSpec::Rails::ControllerExampleGroup
  # ruby2_keywords :method_missing  # <-- Uncomment this line and it will pass
end

DomainObject = Struct.new :string do
  def has_attribute?(encoding:) = string.encoding == encoding
end

RSpec.describe 'some controller', type: :controller do
  it 'whatever' do
    obj = DomainObject.new 'abc'.encode('utf-8')
    expect(obj).to have_attribute encoding: Encoding::UTF_8
  end
end
```

Reply:
Hey, I just changed that first code example. I forgot that bash doesn't understand escaping in single quotes. I added a dollar sign in front of the argument to `ruby -e`, and now it works in bash. Sorry for not verifying that before posting it 😖

Reply:
Yes, I clearly understand why `ruby2_keywords` is needed, and that it is the only approach that works in all Ruby versions that we strive to support. We have it on our roadmap to improve that for all RSpec libraries, but the effort is beyond our current capacity and focus. See https://github.com/rspec/rspec-expectations/issues/1241#issuecomment-749621016

My note was about an entirely different thing. 

The `method_missing` is catching and delegating calls to named routes, and falls back to superclass's implementation otherwise.
But your example is titled "handles dynamic matchers", you define a predicate matcher `has_val?` and `have_val`.
This mismatch caught my eye, as it is confusing.

Do you mind changing the example so that it refers to a named route instead? Does it make sense?

Or was the original problem that you've experienced related to dynamic matchers?
If so, it might be `rspec-core` or `rspec-expectations` that need to be patched instead or in addition to.

Reply:
Alright, I might have misunderstood the whole thing. Your last example clearly indicates the problem.

So, the kwargs error is raised when there's no route, and we delegate to `super`. But when `ruby2_keywords` isn't there, it fails. This case with `super` with no explicit arguments slipped my mind, please accept my apologies.

So this has nothing to do with handling routes specifically.

Let me take a closer look at an earlier time of the day this weekend. Presumably everything is in order.

Reply:
> Alright, I might have misunderstood the whole thing. Your last example clearly indicates the problem.
>
> So, the kwargs error is raised when there's no route, and we delegate to super. But when ruby2_keywords isn't there, it fails. This case with super with no explicit arguments slipped my mind, please accept my apologies.

No worries, it's exceedingly tricky to understand, IME, b/c the spots with the problems can go undetected for a long time and then explode somewhere else completely. And the mental model to understand what the problem is and what the solution is is also very nuanced. I probably put 15 hours of pretty intense analysis into a similar bug in a different library before I actually understood why things weren't working and what to do about it (and I still only partially understand why we're doing any of this 🤷).

> So this has nothing to do with handling routes specifically.

Correct, the `else` clause, which was intended to be a noop, has the unintentional side-effect of changing the invocation signature in Ruby 3.

> Let me take a closer look at an earlier time of the day this weekend. Presumably everything is in order.

👍